Only save callee-saved registers that are used

tyler opened the pull request:
Currently we save all callee-saved registers in the prologue of every function, and restore them in the epilogue. With this PR, we now only save and restore those that are actually used in the function.

I'm not particularly happy with the code in this one. I feel like I'm abusing AllocatableSet. But it appears to work and can be refactored.

Test fixes/changes forthcoming.

sunfishcode commented on lib/cretonne/src/isa/intel/abi.rs:
```diff
@@ -168,6 +169,31 @@ pub fn callee_saved_registers(flags: &shared_settings::Flags) -> &'static [RU] {
+) -> AllocatableSet {
+    let mut all_callee_saved = AllocatableSet::empty();
+    for reg in callee_saved_registers(flags) {
+        all_callee_saved.free(GPR, *reg as RegUnit);
```
While we don't support the Windows x64 calling convention yet, it has some XMM registers as callee-saved. What would you think about renaming `callee_saved_registers` to `callee_saved_gprs`, to remind us to do something else when we need to support callee-saved xmm registers?

Also, it's not new in your patch here, but `callee_saved_registers` doesn't need to be `pub`.

sunfishcode commented on lib/cretonne/src/isa/intel/abi.rs:
```diff
@@ -168,6 +169,31 @@ pub fn callee_saved_registers(flags: &shared_settings::Flags) -> &'static [RU] {
+        match value_loc {
+            &ValueLoc::Reg(ru) => {
+                if !used.is_avail(GPR, ru) {
+                    used.free(GPR, ru);
```
If we can think of a good name, it'd be nice to add a function to `AllocatableSet` aka `RegisterSet` that adds a register to the set without asserting that it's not already set.

sunfishcode replied:
A subtlety here is that the ValueLoc here only contains one register unit of a potentially multi-unit register. This doesn't occur on x86, because we've decided to ignore the H registers in the register allocator. But it wouldn't necessarily be ok for other architectures, so we should put comments here advising people not to blindly reuse this code without thinking about it.

tyler commented on filetests/isa/intel/prologue-epilogue.cton:
```diff
@@ -9,24 +9,14 @@ ebb0:
     return
 }
 
-; check: function %foo(i64 fp [%rbp], i64 csr [%rbx], i64 csr [%r12], i64 csr [%r13], i64 csr [%r14], i64 csr [%r15]) -> i64 fp [%rbp], i64 csr [%rbx], i64 csr [%r12], i64 csr [%r13], i64 csr [%r14], i64 csr [%r15] system_v {
```
I've adjusted this test to match what's being output now, but I'm not totally convinced that it's correct. @sunfishcode, can you look at this and make sure it's sane? I struggled for a bit to convince it to use a callee-saved register in a synthetic test here, but with your new pre and post opt passes, that's increasingly hard to do. :)

sunfishcode replied:
The code looks fine now, and I've also run some non-trivial tests on it and it seems to be working as intended so far. I'll take a closer look at the test here, and I can also help out and add some tests that use callee-saved registers.

tyler commented:
I think I've addressed all the points brought up. Save for the one about adding an unchecked `free` function to `RegisterSet`... basically just because I couldn't think of a clear name for it.

sunfishcode commented:
Looking at testcases, we're no longer adding unused callee-saved registers as arguments and return values. This works for now, however if we add post-regalloc optimizations or analyses in the future, they may need to special-case callee-saved registers to know that they're live. On the other hand, that might not be a big problem, and it reduces clutter to avoid adding arguments that aren't used. So I'm not sure what's best here yet.

sunfishcode commented:
Thinking about this more, we don't currently add callee-saved registers as arguments/returns on calls, so as things stand right now post-regalloc code would have to special-case them anyway. That leads me to think it's fine that we're not adding them all all the time in the prologue/epilogue code.

sunfishcode commented:
Looks good! This likely won't be the last word on callee-saved registers ever ;-), but I think it's a good step right now.
